### DISCUSSION TITLE
JAVA Java Atomic Type에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#97] [Java]
.

**Comment by jimi567** [comment]
java atomic 타입은 멀티 스레드 환경에서 race condition과 같은 문제가 발생하지 않도록, 안전하게 값을 읽고 쓸 수 있도록 해주는 클래스이다.
즉 말 그대로 원자성을 보장하여, 동시성 문제를 해결하는데 도움을 준다.
atomic 타입의 예로는 
AtomicInteger
AtomicReference, 
AtomicIntegerArray,
AtomicIntegerFieldUpdater 등이 있다.

Atomic Type은 CAS( = Compare And Swap) 방식에 기반해 동기화 문제를 해결한다.
( + CAS 방식이란, 변수의 값을 변경하기 전에 기존에 가지고 있던 값이 예상하던 값과 같을 경우에만 새로운 값으로 할당하는 방법이다. )

단, CAS 연산을 사용하기에 추가적인 메모리를 사용해야 하고, 추가적인 연산을 필요로 하므로 일반적인 연산보다 더 많은 리소스가 발생할 수 있음을 고려해야된다.


(PS. 추가적으로 synchronized 랑 volatile 과의 차이 점이나, ABA 문제 등을 공부해보면 좋을 것 같아요. 너무 양이 방대해서 위에 있는 것 까지만..)

**Comment by nayonsoso** [comment]
자바에서는 synchronized 키워드를 사용해서  쓰레드를 동기화 할 수 있습니다.
하지만 임계 영역에 접근하는 다른 쓰레드들을 block 상태로 만들기 때문에 성능 저하가 발생할 수 있습니다.
이를 해결하기 위해 non-blocking 이면서 원자성을 보장할 수 있는 Atomic Type이 고안되었습니다.

Atomic의 핵심 원리는 CAS (Compare And Swap) 알고리즘입니다.
CAS 알고리즘에는 세개의 인자가 필요합니다.
바로 변수의 메모리 주소, 예상되는 변수의 값, 그리고 새롭게 변수에 쓸 값입니다.
먼저, 변수의 메모리 주소를 사용해 값을 읽어옵니다.
이 값이 예상되는 변수 값와 같은지 비교를 하고,
일치한다면 새로 쓸 값으로 변수의 값을 업데이트 합니다.

이를 통해 대기하지 않고 바로 값을 업데이트 할 수 있기 때문에
성능 저하를 최소화하면서 동기화 문제를 해결할 수 있습니다.

**Comment by le2sky** [comment]
https://steady-coding.tistory.com/568

**Comment by GIVEN53** [comment]
멀티스레드
임계영역
원자성
동기화
non-blocking
CAS
메모리 주소, 현재 값, 변경 값
